src/System.Memory/src/System/Text/EncodingExtensions.cs

```diff
@@ -0,0 +1,655 @@
+                Span<byte> remainingBytes = bytes;
+                Encoder encoder = encoding!.GetEncoder();
+                bool isFinalSegment;
+
+                do
+                {
+                    remainingChars.GetFirstSpan(out ReadOnlySpan<char> firstSpan, out SequencePosition next);
+                    isFinalSegment = remainingChars.IsSingleSegment;
+
+                    int bytesWrittenJustNow = encoder.GetBytes(firstSpan, remainingBytes, flush: isFinalSegment);
+                    remainingBytes = remainingBytes.Slice(bytesWrittenJustNow);
+                    remainingChars = remainingChars.Slice(next);
+                } while (!isFinalSegment);
+
+                return originalBytesLength - remainingBytes.Length; // total number of bytes we wrote
+            }
+        }
+
+        /// <summary>
+        /// Encodes the specified <see cref="ReadOnlySequence{Char}"/> into a <see cref="byte"/> array using the specified <see cref="Encoding"/>.
+        /// </summary>
+        /// <param name="encoding">The <see cref="Encoding"/> which represents how the data in <paramref name="chars"/> should be encoded.</param>
+        /// <param name="chars">The <see cref="ReadOnlySequence{Char}"/> to encode to <see langword="byte"/>s.</param>
+        /// <returns>A <see cref="byte"/> array which represents the encoded contents of <paramref name="chars"/>.</returns>
+        /// <exception cref="EncoderFallbackException">Thrown if <paramref name="chars"/> contains data that cannot be encoded and <paramref name="encoding"/> is configured
+        /// to throw an exception when such data is seen.</exception>
+        public static byte[] GetBytes(this Encoding encoding, in ReadOnlySequence<char> chars)
+        {
+            if (encoding is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.encoding);
+            }
+
+            if (chars.IsSingleSegment)
+            {
+                // If the incoming sequence is single-segment, one-shot this.
+
+                ReadOnlySpan<char> span = chars.FirstSpan;
+
+                byte[] retVal = new byte[encoding!.GetByteCount(span)];
+                int actualBytesWritten = encoding.GetBytes(span, retVal);
+
+                Debug.Assert(actualBytesWritten == retVal.Length, "Unexpected number of elements written.");
+
+                return retVal;
+            }
+            else
+            {
+                // If the incoming sequence is multi-segment, create a stateful Encoder
+                // and use it as the workhorse. On the final iteration we'll pass flush=true.
+
+                Encoder encoder = encoding!.GetEncoder();
+
+                // Maintain a list of all the segments we'll need to concat together.
+                // These will be released back to the pool at the end of the method.
+
+                List<(byte[], int)> listOfSegments = new List<(byte[], int)>(8); // guess the initial capacity
```

Comment:
What led to a guess of 8?

Reply:
I was optimizing for 32 KB of data in 4 KB chunks. After talking with David it seems that guess was a bit too high. I'll let the default ctor do its thing.